### PR TITLE
feat(test): support per-project webServer configuration

### LINKED ISSUE #22496: [Feature] Configure web servers per project
I would love to have the possibility to configure local web servers per project.
In my use case, I have a full suite that runs against a local build of my application and a mock server, and a subset of those tests are used as smoke tests against production.

The problem is that playwright always executes the webservers configured when I run only the `smoke-tests` project.

If I had the possibility to configure the `webServer`s per project, I would be able to avoid waiting for the local applications to be instantiated, like:
```ts
export default defineConfig({
  projects: [
    {
      name: 'functional-chromium',
      grepInvert: /@smoke/,
      use: { ...devices['Desktop Chrome'], baseURL: 'http://localhost:3000' },
      webServer: [
        {
           command: 'yarn serve',
           url: 'http://localhost:3000',
           timeout: 380 * 1000,
           reuseExistingServer: true,
        },
        {
           command: 'yarn mock-server',
           port: 3001,
           reuseExistingServer: true,
        },
      ],
    },
    {
       name: 'smoke-chromium',
       grep: /@smoke/,
       use: { ...devices['Desktop Chrome'], baseURL: 'https://production.app.com' },
    },
  ],
});
```

### PR DESCRIPTION
**Pull request by yury-s**
## Summary
- Add `webServer` to `TestProject` so each project can declare its own dev server(s); a per-project server is launched only when the project is selected (directly via `--project` or through the dependency closure), in addition to any top-level `webServer`.

Fixes https://github.com/microsoft/playwright/issues/22496

### CODE REVIEW

File: packages/playwright/src/plugins/index.ts [hunk 1]
```diff
@@ -30,6 +30,9 @@ export interface TestRunnerPlugin {
 export type TestRunnerPluginRegistration = {
   factory: TestRunnerPlugin | (() => TestRunnerPlugin | Promise<TestRunnerPlugin>);
   instance?: TestRunnerPlugin;
+  // When set, the plugin is only set up when one of these projects (or their
+  // transitive closure of dependencies/teardowns) is selected to run.
+  projectIds?: Set<string>;
```
Comment (pavelfeldman) [inline packages/playwright/src/plugins/index.ts]: projectId

File: packages/playwright/src/runner/tasks.ts [hunk 2]
```diff
@@ -95,12 +95,28 @@ export class TestRun {
...
     this.config = config;
     this.options = options ?? {};
     this.reporter = reporter;
     this.filteredProjects = filterProjects(config.projects, this.options.projectFilter);
+    this.projectClosureIds = new Set();
+    for (const project of buildProjectsClosure(this.filteredProjects).keys())
+      this.projectClosureIds.add(project.id);
+  }
+
+  activePlugins(): TestRunnerPluginRegistration[] {
```
Comment (pavelfeldman) [inline packages/playwright/src/runner/tasks.ts]: Let's compare inside the plugin task in tasks.ts

File: packages/playwright/src/runner/testRunner.ts [hunk 3]
```diff
@@ -340,7 +342,7 @@ export class TestRunner extends EventEmitter<TestRunnerEventMap> {
     const tasks = [
       createApplyRebaselinesTask(),
       createLoadTask('out-of-process', { filterOnly: true, failOnLoadErrors: !!params.failOnLoadErrors, doNotRunDepsOutsideProjectFilter: params.doNotRunDepsOutsideProjectFilter }),
-      ...createRunTestsTasks(config),
+      ...createRunTestsTasks(config, testRun),
```
Comment (pavelfeldman) [inline packages/playwright/src/runner/testRunner.ts]: no need to plumb here and in other places

**Review by pavelfeldman — APPROVED** [review]

**Comment by Skn0tt** [comment]
I looked at edge cases here, and I've come around to thinking we should revert this before it ships. The basic idea — gate each project's server on the project closure — works fine for `playwright test --project=...`, but it falls apart in a few places that matter, and it bakes in some public-API choices I'd rather not be stuck with.

The big one: **it doesn't actually do anything in UI mode or the VS Code extension.** `runGlobalSetup` builds its `TestRun` without a project filter (`testRunner.ts:170`), so the closure ends up containing every project and all the per-project servers start anyway. `runTests` has the filter, but it never re-runs plugin setup, so switching projects in the UI doesn't change which servers run. So the whole "don't wait for the servers you don't need" pitch only holds on the CLI. That's a confusing split for a feature whose entire reason to exist is skipping servers.

Second: **port-only project servers don't wire up `baseURL`.** Top-level `webServer: { port }` sets `PLAYWRIGHT_TEST_BASE_URL`, but per-project goes through `createWebServerPlugin(config, false)` (`webServerPlugin.ts:270`), so `baseURL` just stays undefined, silently. I think people will copy the top-level shape and expect it to work the same. It doesn't, and there's no error telling them why.

A couple of smaller things while I'm here: sharding starts (and tears down) every selected project's server on every shard, even shards with no tests for that project — same as top-level today, but per-project multiplies it. And `globalSetup` can't see a project's `webServer` at all, since it only lives on the internal config and never makes it onto `FullConfig`/`FullProject`.

The base-URL semantics and what `globalSetup` can see are public surface, and the UI-mode and sharding behaviors are things people will rely on. I'd rather pull this from the release and redesign with these cases as actual requirements — UI/VS Code project selection driving which servers run, and a real answer for `baseURL` on port-only project servers — than ship it and be on the hook for the gaps.
